Firestore/core/test/unit/model/value_util_test.cc

```diff
@@ -247,6 +274,129 @@ TEST_F(ValueUtilTest, Ordering) {
+  Add(comparison_groups, EncodeServerTimestamp(kTimestamp1, absl::nullopt));
+  Add(comparison_groups, EncodeServerTimestamp(kTimestamp2, absl::nullopt));
+
+  // strings
+  Add(comparison_groups, "");
+  Add(comparison_groups, "\001\ud7ff\ue000\uffff");
+  Add(comparison_groups, "(╯°□°）╯︵ ┻━┻");
+  Add(comparison_groups, "a");
+  Add(comparison_groups, std::string("abc\0 def", 8));
+  Add(comparison_groups, "abc def");
+  // latin small letter e + combining acute accent + latin small letter b
+  Add(comparison_groups, "e\u0301b");
+  Add(comparison_groups, "æ");
+  // latin small letter e with acute accent + latin small letter a
+  Add(comparison_groups, "\u00e9a");
+
+  // blobs
+  Add(comparison_groups, BlobValue());
+  Add(comparison_groups, BlobValue(0));
+  Add(comparison_groups, BlobValue(0, 1, 2, 3, 4));
+  Add(comparison_groups, BlobValue(0, 1, 2, 4, 3));
+  Add(comparison_groups, BlobValue(255));
+
+  // resource names
+  Add(comparison_groups,
+      GetLowerBound(google_firestore_v1_Value_reference_value_tag));
+  Add(comparison_groups, RefValue(DbId("p1/d1"), Key("c1/doc1")));
+  Add(comparison_groups, RefValue(DbId("p1/d1"), Key("c1/doc2")));
+  Add(comparison_groups, RefValue(DbId("p1/d1"), Key("c10/doc1")));
+  Add(comparison_groups, RefValue(DbId("p1/d1"), Key("c2/doc1")));
+  Add(comparison_groups, RefValue(DbId("p1/d2"), Key("c1/doc1")));
+  Add(comparison_groups, RefValue(DbId("p2/d1"), Key("c1/doc1")));
+
+  // geo points
+  Add(comparison_groups, GeoPoint(-90, -180));
+  Add(comparison_groups, GeoPoint(-90, 0));
+  Add(comparison_groups, GeoPoint(-90, 180));
+  Add(comparison_groups, GeoPoint(0, -180));
+  Add(comparison_groups, GeoPoint(0, 0));
+  Add(comparison_groups, GeoPoint(0, 180));
+  Add(comparison_groups, GeoPoint(1, -180));
+  Add(comparison_groups, GeoPoint(1, 0));
+  Add(comparison_groups, GeoPoint(1, 180));
+  Add(comparison_groups, GeoPoint(90, -180));
+  Add(comparison_groups, GeoPoint(90, 0));
+  Add(comparison_groups, GeoPoint(90, 180));
+
+  // arrays
+  Add(comparison_groups,
+      GetLowerBound(google_firestore_v1_Value_array_value_tag));
+  Add(comparison_groups, Array("bar"));
+  Add(comparison_groups, Array("foo", 1));
+  Add(comparison_groups, Array("foo", 2));
+  Add(comparison_groups, Array("foo", "0"));
+
+  // objects
+  Add(comparison_groups,
+      GetLowerBound(google_firestore_v1_Value_map_value_tag));
+  Add(comparison_groups, Map("bar", 0));
+  Add(comparison_groups, Map("bar", 0, "foo", 1));
+  Add(comparison_groups, Map("foo", 1));
+  Add(comparison_groups, Map("foo", 2));
+  Add(comparison_groups, Map("foo", "0"));
+  Add(comparison_groups,
+      GetUpperBound(google_firestore_v1_Value_map_value_tag));
+
+  for (size_t i = 0; i < comparison_groups.size(); ++i) {
+    for (size_t j = i; j < comparison_groups.size(); ++j) {
+      VerifyStrictAscending(comparison_groups[i], comparison_groups[j],
+                            /* is_same_value */ i == j);
```

Comment:
Can we just pass in the expected comparison result (like we did before)? I think it would be cleaner.